File: src/jeeves_airframe/eval/harness.py
"""EvalHarness — run pipelines against eval datasets, collect metrics."""

from __future__ import annotations

import json
import statistics
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Iterable

from jeeves_airframe.reward.base import RewardFn
from jeeves_airframe.trajectory.collector import StreamableRunner, TrajectoryCollector
from jeeves_airframe.trajectory.types import Trajectory


@dataclass
class EvalDataset:
    """Simple eval dataset: list of (input, expected, metadata) tuples."""

    examples: list[dict[str, Any]]

    @classmethod
    def from_jsonl(cls, path: str | Path) -> EvalDataset:
        examples = []
        with open(path, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line:
                    examples.append(json.loads(line))
        return cls(examples=examples)

    @classmethod
    def from_list(cls, examples: list[dict[str, Any]]) -> EvalDataset:
        return cls(examples=examples)

    def __len__(self) -> int:
        return len(self.examples)

    def __iter__(self):
        return iter(self.examples)


@dataclass
class EvalResult:
    """Results from evaluating a pipeline on a dataset."""

    trajectories: list[Trajectory]
    reward_stats: dict[str, float] = field(default_factory=dict)
    per_stage_metrics: dict[str, dict[str, float]] = field(default_factory=dict)
    aggregate: dict[str, float] = field(default_factory=dict)

    @classmethod
    def from_trajectories(cls, trajectories: list[Trajectory]) -> EvalResult:
        result = cls(trajectories=trajectories)
        result._compute_stats()
        return result

    def _compute_stats(self) -> None:
        if not self.trajectories:
            return

        # Reward stats
        rewards = [t.total_reward for t in self.trajectories]
        self.reward_stats = _stats(rewards)

        # Per-stage metrics
        stage_data: dict[str, dict[str, list[float]]] = {}
        for t in self.trajectories:
            for step in t.steps:
                name = step.stage_name
                if name not in stage_data:
                    stage_data[name] = {
                        "duration_ms": [],
                        "tokens_in": [],
                        "tokens_out": [],
                        "llm_calls": [],
                        "tool_calls": [],
                        "success_rate": [],
                    }
                sd = stage_data[name]
                tr = step.stage_trace
                sd["duration_ms"].append(tr.duration_ms)
                sd["tokens_in"].append(tr.tokens_in)
                sd["tokens_out"].append(tr.tokens_out)
                sd["llm_calls"].append(tr.llm_calls)
                sd["tool_calls"].append(tr.tool_calls)
                sd["success_rate"].append(1.0 if tr.success else 0.0)

        self.per_stage_metrics = {
            stage: {k: statistics.mean(v) for k, v in metrics.items() if v}
            for stage, metrics in stage_data.items()
        }

        # Aggregate
        agg_durations = [t.aggregate_metrics.get("total_duration_ms", 0) for t in self.trajectories]
        agg_tokens = [
            t.aggregate_metrics.get("total_tokens_in", 0) + t.aggregate_metrics.get("total_tokens_out", 0)
            for t in self.trajectories
        ]
        self.aggregate = {
            "avg_duration_ms": statistics.mean(agg_durations) if agg_durations else 0,
            "avg_total_tokens": statistics.mean(agg_tokens) if agg_tokens else 0,
            "num_examples": len(self.trajectories),
        }


class EvalHarness:
    """Run a pipeline against an eval dataset, collect trajectories and metrics.

    Usage:
        harness = EvalHarness(runner, reward_fn=my_reward)
        result = harness.evaluate(eval_dataset)
        print(result.reward_stats)
        print(result.per_stage_metrics)
    """

    def __init__(self, runner: StreamableRunner, reward_fn: RewardFn | None = None):
        self._runner = runner
        self._collector = TrajectoryCollector(reward_fn=reward_fn)

    def evaluate(
        self,
        dataset: EvalDataset | Iterable[dict[str, Any]],
        *,
        pipeline_name: str | None = None,
        input_key: str = "input",
    ) -> EvalResult:
        trajectories: list[Trajectory] = []
        for example in dataset:
            inp = example.get(input_key, "")
            metadata = {k: v for k, v in example.items() if k != input_key}
            t = self._collector.collect(
                self._runner,
                inp,
                pipeline_name=pipeline_name,
                metadata=metadata,
            )
            trajectories.append(t)
        return EvalResult.from_trajectories(trajectories)


def _stats(values: list[float]) -> dict[str, float]:
    if not values:
        return {}
    return {
        "mean": statistics.mean(values),
        "median": statistics.median(values),
        "stdev": statistics.stdev(values) if len(values) > 1 else 0.0,
        "min": min(values),
        "max": max(values),
    }

File: src/eval/harness.rs
//! EvalHarness — run a pipeline against an eval dataset and collect metrics.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::reward::RewardFn;
use crate::trajectory::{Trajectory, TrajectoryCollector};
use jeeves_core::worker::runner::PipelineRunner;

/// Results from evaluating a pipeline on a dataset.
#[derive(Debug, Clone, Serialize)]
pub struct EvalResult {
    pub reward_stats: RewardStats,
    pub per_stage_metrics: HashMap<String, HashMap<String, f64>>,
    pub aggregate: HashMap<String, f64>,
    #[serde(skip)]
    pub trajectories: Vec<Trajectory>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct RewardStats {
    pub mean: f64,
    pub median: f64,
    pub stdev: f64,
    pub min: f64,
    pub max: f64,
}

/// Run a pipeline against an eval dataset, collect trajectories and metrics.
#[derive(Debug)]
pub struct EvalHarness {
    collector: TrajectoryCollector,
}

impl EvalHarness {
    pub fn new(reward_fn: Option<Box<dyn RewardFn>>) -> Self {
        Self {
            collector: TrajectoryCollector::new(reward_fn, false),
        }
    }

    pub async fn evaluate(
        &self,
        runner: &PipelineRunner,
        examples: &[EvalExample],
    ) -> jeeves_core::Result<EvalResult> {
        let mut trajectories = Vec::with_capacity(examples.len());
        for example in examples {
            let t = self
                .collector
                .collect(runner, &example.input, "eval")
                .await?;
            trajectories.push(t);
        }
        Ok(EvalResult::from_trajectories(trajectories))
    }
}

/// Single eval example.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EvalExample {
    pub input: String,
    #[serde(default)]
    pub expected: Option<String>,
    #[serde(default)]
    pub metadata: serde_json::Value,
}

impl EvalResult {
    pub fn from_trajectories(trajectories: Vec<Trajectory>) -> Self {
        let rewards: Vec<f64> = trajectories.iter().map(|t| t.total_reward()).collect();
        let reward_stats = compute_stats(&rewards);

        // Per-stage metrics
        let mut stage_data: HashMap<String, HashMap<String, Vec<f64>>> = HashMap::new();
        for t in &trajectories {
            for step in &t.steps {
                let entry = stage_data
                    .entry(step.stage_name.clone())
                    .or_default();
                entry
                    .entry("duration_ms".into())
                    .or_default()
                    .push(step.stage_trace.duration_ms as f64);
                entry
                    .entry("tokens_in".into())
                    .or_default()
                    .push(step.stage_trace.tokens_in as f64);
                entry
                    .entry("tokens_out".into())
                    .or_default()
                    .push(step.stage_trace.tokens_out as f64);
                entry
                    .entry("llm_calls".into())
                    .or_default()
                    .push(step.stage_trace.llm_calls as f64);
                entry
                    .entry("tool_calls".into())
                    .or_default()
                    .push(step.stage_trace.tool_calls as f64);
                entry
                    .entry("success_rate".into())
                    .or_default()
                    .push(if step.stage_trace.success { 1.0 } else { 0.0 });
            }
        }
        let per_stage_metrics: HashMap<String, HashMap<String, f64>> = stage_data
            .into_iter()
            .map(|(stage, metrics)| {
                let avgs = metrics
                    .into_iter()
                    .map(|(k, v)| (k, mean(&v)))
                    .collect();
                (stage, avgs)
            })
            .collect();

        // Aggregate
        let mut aggregate = HashMap::new();
        aggregate.insert(
            "num_examples".into(),
            trajectories.len() as f64,
        );

        Self {
            reward_stats,
            per_stage_metrics,
            aggregate,
            trajectories,
        }
    }
}

fn compute_stats(values: &[f64]) -> RewardStats {
    if values.is_empty() {
        return RewardStats::default();
    }
    let n = values.len() as f64;
    let mean_val = values.iter().sum::<f64>() / n;
    let mut sorted = values.to_vec();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let median = if sorted.len() % 2 == 0 {
        (sorted[sorted.len() / 2 - 1] + sorted[sorted.len() / 2]) / 2.0
    } else {
        sorted[sorted.len() / 2]
    };
    let variance = values.iter().map(|v| (v - mean_val).powi(2)).sum::<f64>() / (n - 1.0).max(1.0);
    let stdev = variance.sqrt();

    RewardStats {
        mean: mean_val,
        median,
        stdev,
        min: sorted[0],
        max: sorted[sorted.len() - 1],
    }
}

fn mean(values: &[f64]) -> f64 {
    if values.is_empty() {
        0.0
    } else {
        values.iter().sum::<f64>() / values.len() as f64
    }
}
